stan/math/prim/fun/rank.hpp

```diff
@@ -21,7 +21,9 @@ inline int rank(const C& v, int s) {
   check_range("rank", "v", v.size(), s);
   --s;  // adjust for indexing by one
   return apply_vector_unary<C>::reduce(v, [s](const auto& vec) {
-    return (vec.array() < vec.coeff(s)).template cast<int>().sum();
+    const auto& vec_ref = to_ref(vec);
+
+    return (vec_ref.array() < vec_ref.coeff(s)).template cast<int>().sum();
```

Comment:
Doesn't it evaluate twice though? Once for the left hand side and once for the right?

If I revert those changes I get:
```
Expected: (matrix0_expr2_counter) <= (1), actual: 2 vs 1
```

This is the test code, prettied up a bit

```cpp
#include <test/expressions/expression_test_helpers.hpp>

TEST(ExpressionTestPrim, rank0) {
  /*
  rank(row_vector, int) => int
  */
  auto matrix0 = stan::test::make_arg<Eigen::Matrix<double, 1, Eigen::Dynamic>>(0.4, 1);
  int int1 = 1;
  int matrix0_expr2_counter = 0;
  stan::test::counterOp<double> matrix0_expr2_counter_op(&matrix0_expr2_counter);
  auto matrix0_expr2 = matrix0.segment(0,1).unaryExpr(matrix0_expr2_counter_op);

  auto matrix3 = stan::test::make_arg<Eigen::Matrix<double, 1, Eigen::Dynamic>>(0.4, 1);
  int int4 = 1;

  auto result5 = stan::math::eval(stan::math::rank(matrix0_expr2,int1));
  auto result6 = stan::math::eval(stan::math::rank(matrix3,int4));
  EXPECT_STAN_EQ(result5,result6);
  EXPECT_LEQ_ONE(matrix0_expr2_counter);
}
```

Reply:
Only a single element is evaluated twice. I think that is better than forcing evaluation or everything.

> If I revert those changes I get:

 There used to be a special exception in expression tests for this function. Maybe it should be kept in.

Reply:
This would be a wrong thing to do if the expression was a matrix-vector product.

I could add it to the ignored list but I prefer to just not have the exception. I'm not really sure how this function is used in practice (if it's used in a way where performance is critical or not). I'm assuming not.

Reply:
I see your point. Fine.